### ISSUE TITLE
Many renaming passes are incapable of handling shorthand object literals and destructured objects

### ISSUE THREAD

**Issue by shicks** [google/closure-compiler#2685]
03e49f99daa73be60577eae45db2f75b1fc5eb1a was rolled back because we changed the order of rest parameter rewriting to be *before* block scope rewriting.  It turns out that rewriting rest parameters introduces a shadowed variable name, which was not getting renamed correctly in short-hand object literals.

A simple repro of the problem this causes is:

```js
let x;
{
  let x = 4;
  assertObjectEquals({x: 4}, {x});
}
```

This test currently fails when transpiled, since it is incorrectly rewritten as

```js
let x;
{
  let x$0 = 4;
  assertObjectEquals({x: 4}, {x});
}
```

When the inner-scope `x` is rewritten to `x$0`, it needed to also update the shorthand object literal to `{x: x$0}`, but it did not because renaming is based on `NodeUtil.isReferenceName(n)` which requires `n.isName()`, which is false for the `STRING_KEY` in the shorthand object, and it uses `Node#setString` to do the renaming, which is obviously incorrect as well.

It makes sense to fix `isReferenceName` to return true for child-less `STRING_KEY`s as well, and to add a `NodeUtil#renameReference(Node, String)` method that will handle these names correctly.  Unfortunately, `isReferenceName` is called a number of times and so I'm currently auditing these usages.  And `setString` is also ubiquitous, so there's a lot of work to be done to figure out which of these need to change to `renameReference`.

**Comment by brad4d** [comment]
When reproducing this this did you have normalization turned on in the test case?
I believe normalization is supposed to rewrite `{x} = something` as `{x: x} = something`, and the target `x` will be a `NAME` node.

**Comment by shicks** [comment]
It was pointed out that `Normalize` already rewrites all shorthand properties before any optimizations occur anyway.  Whenever we're lowering from ES6 to ES5, it makes sense to just do this as early as possible, since it needs to happen sometime anyway.  So optimizations already don't need to worry about shorthand properties (though currently destructuring is not normalized, so that will still be an issue).

There are still several outstanding questions:
1. should we also normalize destructuring assignments?
2. given that we already do this unconditionally before optimizations, is it worth just normalizing unconditionally at the very beginning, regardless of whether the output language is ES6 or better, and then at the very end folding shorthands back in if the output language can support it?

@concavelenz I'd be interested to hear your thoughts on these two questions.

**Comment by ChadKillingsworth** [comment]
Having to specifically handle shorthand and destructuring during module rewriting for CommonJS is very cumbersome. I'd very much like to see Normalize take care of this and happen before passes like that if possible.

**Comment by shicks** [comment]
This was closed by aaa02d0.
